From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] opengl32: wow64 pointer getters, glGetUniformOffsetEXT, wglAllocateMemoryNV, unthunkable extensions withheld (3635)

opengl32: wow64 pointer getters, glGetUniformOffsetEXT, wglAllocateMemoryNV, unthunkable extensions withheld

unix_thunks.c is regenerated with make_opengl (scratch copy, cache in work/oglgen; wgl_driver.h untouched). Gate fastb-gapogl (64- and 32-bit probe, DSA getters exercised on llvmpipe). Mutants: SG_MUTANT_WOW64_EXT (filter) and two thunk edits (glGetPointerv write, glGetPointerIndexedvEXT write) shown caught. opengl32 conformance 64/32: 0 failures as stock. Unverified on hardware: glGetTexParameterPointervAPPLE, glGetVariantPointervEXT, glGetVertexAttribPointervNV, glGetUniformOffsetEXT, wglAllocateMemoryNV (extensions not exposed by llvmpipe). CORE review: glGetPointerv fix. 
---
diff --git a/dlls/opengl32/make_opengl b/dlls/opengl32/make_opengl
index 1e55fdf..b8d15e1 100755
--- a/dlls/opengl32/make_opengl
+++ b/dlls/opengl32/make_opengl
@@ -216,6 +216,7 @@ my %manual_wow64_thunks =
      "glUnmapNamedBuffer" => 1,
      "glUnmapNamedBufferEXT" => 1,
      "glWaitSync" => 1,
+     "wglAllocateMemoryNV" => 1,
      "wglCreateContext" => 1,
      "wglCreateContextAttribsARB" => 1,
      "wglCreatePbufferARB" => 1,
@@ -254,8 +255,21 @@ my %pointer_is_offset =
     (
      "glGetPointerv" => "params",
      "glGetPointervEXT" => "params",
+     "glGetPointerIndexedvEXT" => "data",
+     "glGetPointeri_vEXT" => "params",
+     "glGetTexParameterPointervAPPLE" => "params",
+     "glGetVariantPointervEXT" => "data",
+     "glGetVertexArrayPointeri_vEXT" => "param",
+     "glGetVertexArrayPointervEXT" => "param",
      "glGetVertexAttribPointerv" => "pointer",
      "glGetVertexAttribPointervARB" => "pointer",
+     "glGetVertexAttribPointervNV" => "pointer",
+    );
+# functions returning an integer offset typed as a pointer-sized integer: the
+# offset always fits 32 bits, so the result is passed through
+my %wow64_ret_is_offset =
+    (
+     "glGetUniformOffsetEXT" => 1,
     );
 
 #
@@ -466,7 +480,7 @@ sub generate_wow64_thunk($$$)
     }
     $ret .= "    };\n";
 
-    if ($need_manual_thunk || get_wow64_arg_type( $func->[0] ) =~ /PTR32/)
+    if ($need_manual_thunk || (get_wow64_arg_type( $func->[0] ) =~ /PTR32/ && !defined $wow64_ret_is_offset{$name}))
     {
         $ret .= "    FIXME( \"params32 %p, params %p stub!\\n\", params32, &params );\n";
         $ret .= "    return STATUS_NOT_IMPLEMENTED;\n";
@@ -476,6 +490,15 @@ sub generate_wow64_thunk($$$)
 
     $ret .= "    NTSTATUS status;\n";
 
+    foreach my $arg (@{$func->[1]})
+    {
+        next unless defined $pointer_is_offset{$_};
+        my $pname = get_arg_name( $arg );
+        next unless $pname =~ $pointer_is_offset{$_};
+        $ret .= "    void *$pname\_out = params32->$pname ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->$pname )) : NULL;\n";
+        $ret .= "    params.$pname = &$pname\_out;\n";
+    }
+
     foreach my $arg (@{$func->[1]})
     {
         next unless $arg->textContent() =~ /\*.*\*/ || $arg->textContent() =~ /(sizei|int)ptr.*\*/;
@@ -505,7 +528,7 @@ sub generate_wow64_thunk($$$)
         next unless defined $pointer_is_offset{$_};
         my $pname = get_arg_name( $arg );
         next unless $pname =~ $pointer_is_offset{$_};
-        $ret .= "    params32->$pname = PtrToUlong( params.$pname );\n";
+        $ret .= "    if (params32->$pname) *(PTR32 *)ULongToPtr( params32->$pname ) = PtrToUlong( $pname\_out );\n";
     }
 
     $ret .= "    return status;\n";
diff --git a/dlls/opengl32/unix_thunks.c b/dlls/opengl32/unix_thunks.c
index 5afe6c3..36f043e 100644
--- a/dlls/opengl32/unix_thunks.c
+++ b/dlls/opengl32/unix_thunks.c
@@ -24068,7 +24068,7 @@ static NTSTATUS ext_glWriteMaskEXT( void *args )
     return STATUS_SUCCESS;
 }
 
-static NTSTATUS ext_wglAllocateMemoryNV( void *args )
+NTSTATUS ext_wglAllocateMemoryNV( void *args )
 {
     struct wglAllocateMemoryNV_params *params = args;
     const struct opengl_funcs *funcs = params->teb->glTable;
@@ -29628,8 +29628,10 @@ static NTSTATUS wow64_gl_glGetPointerv( void *args )
         .pname = params32->pname,
     };
     NTSTATUS status;
+    void *params_out = params32->params ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->params )) : NULL;
+    params.params = &params_out;
     status = gl_glGetPointerv( &params );
-    params32->params = PtrToUlong( params.params );
+    if (params32->params) *(PTR32 *)ULongToPtr( params32->params ) = PtrToUlong( params_out );
     return status;
 }
 
@@ -51955,8 +51957,12 @@ static NTSTATUS wow64_ext_glGetPointerIndexedvEXT( void *args )
         .target = params32->target,
         .index = params32->index,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *data_out = params32->data ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->data )) : NULL;
+    params.data = &data_out;
+    status = ext_glGetPointerIndexedvEXT( &params );
+    if (params32->data) *(PTR32 *)ULongToPtr( params32->data ) = PtrToUlong( data_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetPointeri_vEXT( void *args )
@@ -51974,8 +51980,12 @@ static NTSTATUS wow64_ext_glGetPointeri_vEXT( void *args )
         .pname = params32->pname,
         .index = params32->index,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *params_out = params32->params ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->params )) : NULL;
+    params.params = &params_out;
+    status = ext_glGetPointeri_vEXT( &params );
+    if (params32->params) *(PTR32 *)ULongToPtr( params32->params ) = PtrToUlong( params_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetPointervEXT( void *args )
@@ -51992,8 +52002,10 @@ static NTSTATUS wow64_ext_glGetPointervEXT( void *args )
         .pname = params32->pname,
     };
     NTSTATUS status;
+    void *params_out = params32->params ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->params )) : NULL;
+    params.params = &params_out;
     status = ext_glGetPointervEXT( &params );
-    params32->params = PtrToUlong( params.params );
+    if (params32->params) *(PTR32 *)ULongToPtr( params32->params ) = PtrToUlong( params_out );
     return status;
 }
 
@@ -53669,8 +53681,12 @@ static NTSTATUS wow64_ext_glGetTexParameterPointervAPPLE( void *args )
         .target = params32->target,
         .pname = params32->pname,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *params_out = params32->params ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->params )) : NULL;
+    params.params = &params_out;
+    status = ext_glGetTexParameterPointervAPPLE( &params );
+    if (params32->params) *(PTR32 *)ULongToPtr( params32->params ) = PtrToUlong( params_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetTexParameterxvOES( void *args )
@@ -54431,8 +54447,10 @@ static NTSTATUS wow64_ext_glGetUniformOffsetEXT( void *args )
         .program = params32->program,
         .location = params32->location,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    status = ext_glGetUniformOffsetEXT( &params );
+    params32->ret = (UINT_PTR)params.ret;
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetUniformSubroutineuiv( void *args )
@@ -54847,8 +54865,12 @@ static NTSTATUS wow64_ext_glGetVariantPointervEXT( void *args )
         .id = params32->id,
         .value = params32->value,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *data_out = params32->data ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->data )) : NULL;
+    params.data = &data_out;
+    status = ext_glGetVariantPointervEXT( &params );
+    if (params32->data) *(PTR32 *)ULongToPtr( params32->data ) = PtrToUlong( data_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetVaryingLocationNV( void *args )
@@ -54979,8 +55001,12 @@ static NTSTATUS wow64_ext_glGetVertexArrayPointeri_vEXT( void *args )
         .index = params32->index,
         .pname = params32->pname,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *param_out = params32->param ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->param )) : NULL;
+    params.param = &param_out;
+    status = ext_glGetVertexArrayPointeri_vEXT( &params );
+    if (params32->param) *(PTR32 *)ULongToPtr( params32->param ) = PtrToUlong( param_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetVertexArrayPointervEXT( void *args )
@@ -54998,8 +55024,12 @@ static NTSTATUS wow64_ext_glGetVertexArrayPointervEXT( void *args )
         .vaobj = params32->vaobj,
         .pname = params32->pname,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *param_out = params32->param ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->param )) : NULL;
+    params.param = &param_out;
+    status = ext_glGetVertexArrayPointervEXT( &params );
+    if (params32->param) *(PTR32 *)ULongToPtr( params32->param ) = PtrToUlong( param_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetVertexArrayiv( void *args )
@@ -55270,8 +55300,10 @@ static NTSTATUS wow64_ext_glGetVertexAttribPointerv( void *args )
         .pname = params32->pname,
     };
     NTSTATUS status;
+    void *pointer_out = params32->pointer ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->pointer )) : NULL;
+    params.pointer = &pointer_out;
     status = ext_glGetVertexAttribPointerv( &params );
-    params32->pointer = PtrToUlong( params.pointer );
+    if (params32->pointer) *(PTR32 *)ULongToPtr( params32->pointer ) = PtrToUlong( pointer_out );
     return status;
 }
 
@@ -55291,8 +55323,10 @@ static NTSTATUS wow64_ext_glGetVertexAttribPointervARB( void *args )
         .pname = params32->pname,
     };
     NTSTATUS status;
+    void *pointer_out = params32->pointer ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->pointer )) : NULL;
+    params.pointer = &pointer_out;
     status = ext_glGetVertexAttribPointervARB( &params );
-    params32->pointer = PtrToUlong( params.pointer );
+    if (params32->pointer) *(PTR32 *)ULongToPtr( params32->pointer ) = PtrToUlong( pointer_out );
     return status;
 }
 
@@ -55311,8 +55345,12 @@ static NTSTATUS wow64_ext_glGetVertexAttribPointervNV( void *args )
         .index = params32->index,
         .pname = params32->pname,
     };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS status;
+    void *pointer_out = params32->pointer ? ULongToPtr( *(PTR32 *)ULongToPtr( params32->pointer )) : NULL;
+    params.pointer = &pointer_out;
+    status = ext_glGetVertexAttribPointervNV( &params );
+    if (params32->pointer) *(PTR32 *)ULongToPtr( params32->pointer ) = PtrToUlong( pointer_out );
+    return status;
 }
 
 static NTSTATUS wow64_ext_glGetVertexAttribdv( void *args )
@@ -91886,29 +91924,6 @@ static NTSTATUS wow64_ext_glWriteMaskEXT( void *args )
     return status;
 }
 
-static NTSTATUS wow64_ext_wglAllocateMemoryNV( void *args )
-{
-    struct
-    {
-        PTR32 teb;
-        GLsizei size;
-        GLfloat readfreq;
-        GLfloat writefreq;
-        GLfloat priority;
-        PTR32 ret;
-    } *params32 = args;
-    struct wglAllocateMemoryNV_params params =
-    {
-        .teb = get_teb64(params32->teb),
-        .size = params32->size,
-        .readfreq = params32->readfreq,
-        .writefreq = params32->writefreq,
-        .priority = params32->priority,
-    };
-    FIXME( "params32 %p, params %p stub!\n", params32, &params );
-    return STATUS_NOT_IMPLEMENTED;
-}
-
 static NTSTATUS wow64_ext_wglBindTexImageARB( void *args )
 {
     struct
@@ -92285,6 +92300,7 @@ extern NTSTATUS wow64_ext_glUnmapBufferARB( void *args );
 extern NTSTATUS wow64_ext_glUnmapNamedBuffer( void *args );
 extern NTSTATUS wow64_ext_glUnmapNamedBufferEXT( void *args );
 extern NTSTATUS wow64_ext_glWaitSync( void *args );
+extern NTSTATUS wow64_ext_wglAllocateMemoryNV( void *args );
 extern NTSTATUS wow64_ext_wglCreateContextAttribsARB( void *args );
 extern NTSTATUS wow64_ext_wglCreatePbufferARB( void *args );
 extern NTSTATUS wow64_ext_wglGetExtensionsStringARB( void *args );
diff --git a/dlls/opengl32/unix_wgl.c b/dlls/opengl32/unix_wgl.c
index 6e6fcdf..b6fb42e 100644
--- a/dlls/opengl32/unix_wgl.c
+++ b/dlls/opengl32/unix_wgl.c
@@ -159,6 +159,30 @@ static BOOL has_extension( const char *list, const char *ext, size_t len )
     return FALSE;
 }
 
+/* Extensions whose entry points take or return host pointers that cannot be
+ * represented for a 32-bit client (client-side pointer lists read at draw
+ * time, driver-mapped memory, OpenCL / X11 / Vulkan interop handles). The wow64
+ * thunks for them are not implemented, so they are not advertised, as with
+ * GL_ARB_buffer_storage. */
+static BOOL is_wow64_unsupported_extension( const char *ext )
+{
+    static const char *const names[] =
+    {
+        "GL_ARB_buffer_storage", "GL_EXT_buffer_storage",
+        "GL_IBM_vertex_array_lists", "GL_INTEL_parallel_arrays", "GL_SUN_triangle_list",
+        "GL_ATI_map_object_buffer", "GL_INTEL_map_texture", "GL_ARB_cl_event",
+        "GL_EXT_x11_sync_object", "GL_NV_draw_vulkan_image",
+    };
+    unsigned int i;
+
+#ifdef SG_MUTANT_WOW64_EXT
+    return FALSE;
+#endif
+    if (!is_win64 || !is_wow64()) return FALSE;
+    for (i = 0; i < ARRAY_SIZE(names); i++) if (!strcmp( ext, names[i] )) return TRUE;
+    return FALSE;
+}
+
 static GLubyte *filter_extensions_list( const char *extensions, const char *disabled )
 {
     const char *end;
@@ -179,8 +203,8 @@ static GLubyte *filter_extensions_list( const char *extensions, const char *disa
         p[end - extensions] = 0;
 
         /* We do not support GL_MAP_PERSISTENT_BIT, and hence
-         * ARB_buffer_storage, on wow64. */
-        if (is_win64 && is_wow64() && (!strcmp( p, "GL_ARB_buffer_storage" ) || !strcmp( p, "GL_EXT_buffer_storage" )))
+         * ARB_buffer_storage, on wow64 (see is_wow64_unsupported_extension). */
+        if (is_wow64_unsupported_extension( p ))
         {
             TRACE( "-- %s (disabled due to wow64)\n", p );
         }
@@ -250,8 +274,8 @@ static GLuint *filter_extensions_index( TEB *teb, const char *disabled )
         ext = (const char *)funcs->ext.p_glGetStringi( GL_EXTENSIONS, j );
 
         /* We do not support GL_MAP_PERSISTENT_BIT, and hence
-         * ARB_buffer_storage, on wow64. */
-        if (is_win64 && is_wow64() && (!strcmp( ext, "GL_ARB_buffer_storage" ) || !strcmp( ext, "GL_EXT_buffer_storage" )))
+         * ARB_buffer_storage, on wow64 (see is_wow64_unsupported_extension). */
+        if (is_wow64_unsupported_extension( ext ))
         {
             TRACE( "-- %s (disabled due to wow64)\n", ext );
             disabled_index[i++] = j;
@@ -2286,4 +2310,33 @@ NTSTATUS wow64_get_pixel_formats( void *args )
     return status;
 }
 
+/* WGL_NV_vertex_array_range: the memory comes from the driver and has to be
+ * addressable by the 32-bit client. When it is not, the allocation fails (NULL,
+ * the documented failure result) and the client falls back to plain vertex arrays. */
+NTSTATUS wow64_ext_wglAllocateMemoryNV( void *args )
+{
+    struct
+    {
+        PTR32 teb;
+        GLsizei size;
+        GLfloat readfreq;
+        GLfloat writefreq;
+        GLfloat priority;
+        PTR32 ret;
+    } *params32 = args;
+    TEB *teb = get_teb64( params32->teb );
+    const struct opengl_funcs *funcs = teb->glTable;
+    void *ptr;
+
+    ptr = funcs->ext.p_wglAllocateMemoryNV( params32->size, params32->readfreq, params32->writefreq,
+                                            params32->priority );
+    if (ptr && ULongToPtr( PtrToUlong( ptr )) != ptr)
+    {
+        funcs->ext.p_wglFreeMemoryNV( ptr );
+        ptr = NULL;
+    }
+    params32->ret = PtrToUlong( ptr );
+    return STATUS_SUCCESS;
+}
+
 #endif
